src/hotspot/share/gc/g1/g1FullGCPrepareTask.cpp

```diff
@@ -191,6 +205,50 @@ void G1FullGCPrepareTask::G1CalculatePointersClosure::prepare_for_compaction(Hea
   prepare_for_compaction_work(_cp, hr);
 }
 
+void G1FullGCPrepareTask::G1CalculatePointersClosure::prepare_for_skipping_compaction(HeapRegion* hr) {
+  HeapRegion* current = hr;
+  HeapWord* limit = current->top();
+  HeapWord* next_addr = current->bottom();
+  HeapWord* live_end = current->bottom();
+  _skipping_compaction_set->append(current);
+  HeapWord* threshold = current->initialize_threshold();
+  HeapWord* pre_addr;
+
+  while (next_addr < limit) {
+    Prefetch::write(next_addr, PrefetchScanIntervalInBytes);
+    pre_addr = next_addr;
+
+    if (_bitmap->is_marked(next_addr)) {
+      oop obj = oop(next_addr);
+      size_t obj_size = obj->size();
+      // Object should not move but mark-word is used so it looks like the
+      // object is forwarded. Need to clear the mark and it's no problem
+      // since it will be restored by preserved marks. There is an exception
+      // with BiasedLocking, in this case forwardee() will return NULL
+      // even if the mark-word is used. This is no problem since
+      // forwardee() will return NULL in the compaction phase as well.
+      if (obj->forwardee() != NULL) {
+        obj->init_mark();
+      }
+
+      next_addr += obj_size;
+      // update live byte range end
+      live_end = next_addr;
+    } else {
+      next_addr = _bitmap->get_next_marked_addr(next_addr, limit);
+      assert(next_addr > live_end, "next_addr must be bigger than live_end");
+      assert(next_addr == limit || _bitmap->is_marked(next_addr), "next_addr is the limit or is marked");
+      // fill dummy object to replace dead range
+      Universe::heap()->fill_with_dummy_object(live_end, next_addr, true);
+    }
+
+    if (next_addr > threshold) {
+      threshold = current->cross_threshold(pre_addr, next_addr);
+    }
+  }
+  assert(next_addr == limit, "Should stop the scan at the limit.");
+}
+
```

Comment:
I think all this code is not required/duplicate of the existing code that handles pinned regions (added in https://bugs.openjdk.java.net/browse/JDK-8253600). Please have a look and see if that code can be reused/repurposed.

I mean, these "skipped regions" should be equivalent to "pinned" regions wrt to required object handling. I admit I haven't actually looked how to do this, but this seems awfully similar with the only difference that skipped regions are determined only after marking and pinned regions sometimes before (full) gc.

The other handling should be the same. So *at worst* temporarily marking these regions as "pinned" should be sufficient for this functionality to be able to reuse the mechanism.

Reply:
After further investigation, I think so too.